### PR TITLE
Handle capabilities prefixes

### PR DESCRIPTION
**Pull request by dpgraham**
* Strip out 'appium:' prefixes
* Reject standard capabilities that contain the 'appium:' prefix
* Expose 'processCapabilities' as API method

(see https://www.w3.org/TR/webdriver/#dfn-extension-capabilities for reference)

(related to https://github.com/appium/appium/issues/9785)

### CODE REVIEW

File: lib/basedriver/capabilities.js [hunk 1]
```diff
@@ -51,6 +52,47 @@ function validateCaps (caps, constraints = {}, skipPresenceConstraint = false) {
   return caps;
 }
 
+// Standard, non-prefixed capabilities (see https://www.w3.org/TR/webdriver/#dfn-table-of-standard-capabilities)
+const STANDARD_CAPS = ['browserName', 'browserVersion', 'platformName', 'acceptInsecureCerts', 'pageLoadStrategy', 'proxy', 'setWindowRect', 'timeouts', 'unhandledPromptBehavior'];
```
Comment (mykola-mokhnach) [inline lib/basedriver/capabilities.js]: I'd rather format these as a column

File: lib/basedriver/capabilities.js [hunk 2]
```diff
@@ -51,6 +52,47 @@ function validateCaps (caps, constraints = {}, skipPresenceConstraint = false) {
...
+
+function isStandardCap (cap) {
+  return !!_.find(STANDARD_CAPS, (standardCap) => standardCap.toLowerCase() === cap.toLowerCase());
```
Comment (mykola-mokhnach) [inline lib/basedriver/capabilities.js]: `${cap}` would be safer

File: lib/basedriver/capabilities.js [hunk 3]
```diff
@@ -51,6 +52,47 @@ function validateCaps (caps, constraints = {}, skipPresenceConstraint = false) {
...
+}
+
+// If the 'appium:' prefix was provided and it's a valid capability, strip out the prefix (see https://www.w3.org/TR/webdriver/#dfn-extension-capabilities)
+function stripAppiumPrefixes (caps) {
+  const prefix = 'appium:';
+  const prefixedCaps = _.filter(_.keys(caps), cap => cap.startsWith(prefix));
```
Comment (mykola-mokhnach) [inline lib/basedriver/capabilities.js]: `${cap}` would be safer

File: lib/basedriver/capabilities.js [hunk 4]
```diff
@@ -51,6 +52,47 @@ function validateCaps (caps, constraints = {}, skipPresenceConstraint = false) {
...
+  const badPrefixedCaps = [];
+  const unprefixedCaps = _.filter(_.keys(caps), (cap) => (
```
Comment (mykola-mokhnach) [inline lib/basedriver/capabilities.js]: same as above

File: lib/basedriver/capabilities.js [hunk 5]
```diff
@@ -51,6 +52,47 @@ function validateCaps (caps, constraints = {}, skipPresenceConstraint = false) {
...
+    cap.indexOf(':') < 0 && !isStandardCap(cap)
```
Comment (mykola-mokhnach) [inline lib/basedriver/capabilities.js]: `.includes` might be more readable

File: lib/basedriver/capabilities.js (hunk 2 above)
Comment (dpgraham) [reply lib/basedriver/capabilities.js]: Sorry, I don't think I understand. What does `${cap}` replace?

File: lib/basedriver/capabilities.js [hunk 6]
```diff
@@ -51,6 +52,47 @@ function validateCaps (caps, constraints = {}, skipPresenceConstraint = false) {
...
 
+// Standard, non-prefixed capabilities (see https://www.w3.org/TR/webdriver/#dfn-table-of-standard-capabilities)
+const STANDARD_CAPS = ['browserName', 'browserVersion', 'platformName', 'acceptInsecureCerts', 'pageLoadStrategy', 'proxy', 'setWindowRect', 'timeouts', 'unhandledPromptBehavior'];
+
+function isStandardCap (cap) {
+  return !!_.find(STANDARD_CAPS, (standardCap) => standardCap.toLowerCase() === cap.toLowerCase());
+}
+
+// If the 'appium:' prefix was provided and it's a valid capability, strip out the prefix (see https://www.w3.org/TR/webdriver/#dfn-extension-capabilities)
+function stripAppiumPrefixes (caps) {
```
Comment (mykola-mokhnach) [inline lib/basedriver/capabilities.js]: I usually add a comment if a function mutates its argument(s).

File: test/basedriver/capabilities-specs.js [hunk 7]
```diff
@@ -214,5 +215,19 @@ describe('caps', () => {
         firstMatch: [{foo: 'bar'}]
       }).should.deep.equal({hello: 'world', foo: 'bar'});
     });
+
+    it('should strip out the "appium:" prefix for non-standard capabilities', () => {
```
Comment (mykola-mokhnach) [inline test/basedriver/capabilities-specs.js]: mocha prefers functions over lambdas

File: lib/basedriver/capabilities.js (hunk 2 above)
Comment (mykola-mokhnach) [reply lib/basedriver/capabilities.js]: I mean that it is not guaranteed that `cap` is of type string, which may cause an unexpected exception.

**Comment by dpgraham** [comment]
Fixes pushed.

File: lib/basedriver/capabilities.js (hunk 3 above)
Comment (mykola-mokhnach) [reply lib/basedriver/capabilities.js]: please fix `cap` while using it with `startsWith`

File: lib/basedriver/capabilities.js [hunk 8]
```diff
@@ -53,19 +53,30 @@ function validateCaps (caps, constraints = {}, skipPresenceConstraint = false) {
...
 
 // If the 'appium:' prefix was provided and it's a valid capability, strip out the prefix (see https://www.w3.org/TR/webdriver/#dfn-extension-capabilities)
+// (NOTE: Method is destructive and mutates contents of caps)
 function stripAppiumPrefixes (caps) {
   const prefix = 'appium:';
   const prefixedCaps = _.filter(_.keys(caps), cap => cap.startsWith(prefix));
   const badPrefixedCaps = [];
   const unprefixedCaps = _.filter(_.keys(caps), (cap) => (
-    cap.indexOf(':') < 0 && !isStandardCap(cap)
+    cap.includes(':') && !isStandardCap(`${cap}`)
```
Comment (mykola-mokhnach) [inline lib/basedriver/capabilities.js]: it is not necessary to do it here if conversation to string is already done in `isStandardCap`

File: lib/basedriver/capabilities.js [hunk 9]
```diff
@@ -51,6 +52,58 @@ function validateCaps (caps, constraints = {}, skipPresenceConstraint = false) {
...
+
+function isStandardCap (cap) {
+  return !!_.find(STANDARD_CAPS, (standardCap) => standardCap.toLowerCase() === `${cap}`.toLowerCase());
+}
+
+// If the 'appium:' prefix was provided and it's a valid capability, strip out the prefix (see https://www.w3.org/TR/webdriver/#dfn-extension-capabilities)
+// (NOTE: Method is destructive and mutates contents of caps)
+function stripAppiumPrefixes (caps) {
+  const prefix = 'appium:';
+  const prefixedCaps = _.filter(_.keys(caps), cap => cap.startsWith(prefix));
```
Comment (mykola-mokhnach) [inline lib/basedriver/capabilities.js]:
```
const prefixedCaps = _.filter(_.keys(caps), cap => `${cap}`.startsWith(prefix));
```

File: lib/basedriver/capabilities.js [hunk 10]
```diff
@@ -51,6 +52,58 @@ function validateCaps (caps, constraints = {}, skipPresenceConstraint = false) {
...
+  const badPrefixedCaps = [];
+  const unprefixedCaps = _.filter(_.keys(caps), (cap) => (
+    cap.includes(':') && !isStandardCap(cap)
```
Comment (mykola-mokhnach) [inline lib/basedriver/capabilities.js]:
```
`${cap}`.includes(':') && !isStandardCap(cap)
```

**Pull request merged**